Add-ons manager: It is now possible to disable individual add-ons. re #3090

josephsl opened the pull request:
It is now possible to disable individual add-ons from add-ons manager. To support this, add-on handler has been modified as follows:
- Added a new set in add-on state that records names of add-ons that are disabled.
- Added a property in add-on bundle data structure that returns if the add-on is disabled. Also added several pending properties (pendingEnable, pendingDisable).
- Added disable/enable button in add-ons manager to enable/disable add-ons. This works regardless of whether NVDA has loaded add-ons or not.
- Added explanation of this ability in the user guide.

josephsl committed:
Add-ons subsystem: Base routines to disable add-ons upon request. re #3090.

In order to allow users to disable add-ons from add-ons manager or from command line, it is necessary to tweak the add-ons subsystem a bit.
Specifically:
* Core.py: If all add-ons are disabled, this fact is recorded in the log, useful for troubleshooting purposes.
* addonHandler.py: Added a new pending disable set so add-ons can be disabled upon startup. The user interface to manipulate this set will be committed later. Also added a function (disableAddonsIfAny) that'll save current disable state to a private set (can be extended to supporting adding add-ons from command line).
* addonHandler.Addon: Added isDisabled property to report if the currently selected add-on is disabled. Also modified isRunning attribute to check if the add-on is disabled. These properties will come in handy when presenting state information in add-ons manager.
So far, tests show it is working. Next on the list is adding few more properties such as pending enable/disable to addonHandler.Addon, then start implementing the user interface.

josephsl committed:
Add-ons manager and add-ons subsystem: NVDA can disable certain add-ons upon request (part 1). re #3090.

Added a routine in addonHandler.Addon to enable or disable add-ons upon request, set via a new button in add-ons manager (no hotkey assigned to prevent users from accidentally disabling wrong add-ons). The next commit will be display strings for when an add-on is enable and disabled.

josephsl committed:
Add-ons subsystem and UI: present a warning before disabling the selected add-on. re #3090.

Suggested by Norlia Martinez: prompt for confirmation before disabling the selected add-on to make sure users know what they are doing.
Also caught KeyError exception, likely to be encountered when installing a build which includes this feature for the first time.

josephsl committed:
Add-ons subsystem, UI and user guide: enable/disable status are now shown, documented add-on enable/disable feature. re #3090.

Although it was suggested to use a dedicated set for pending enables, it is better to use a module-level set, as pendingDisableSet is really intended to tell disabled add-ons set what to disable. Also, because isDisabled flag takes precedence, enable/disalbe status is fetched early.
Also documented add-on enable/disable feature in the user guide. As of this commit, all the pieces for enabling/disabling add-ons and its user interface are there (next up is command-line version, a bit tricky but doable).

josephsl committed:
Merge branch 'master' into i3090

josephsl committed:
Merge branch 'master' into i3090

josephsl committed:
Remove a comment on allowing add-ons to be disabled from command line, as it is not feasible. re #3090

josephsl committed:
Add-ons manager: Changed UI string for diable add-on dialog and restart screen. re #3090
Good catch by Derek Riemer: always strive for consistency in user interface strings.

michaelDCurran commented on source/gui/addonGui.py:
```diff
@@ -206,18 +221,26 @@ def refreshAddonsList(self,activeIndex=0):
 			self.removeButton.Disable()
 
+	def _shouldDisable(self, addon):
+		return not (addon.isPendingDisable or (addon.isDisabled and not addon.isPendingEnable and not addon.isPendingDisable))
```
This is a confusing bit of logic. Do you need to check addon.isPendingDisable twice? I would have thought that if it got to the inner parenthesis, isPendingDisable could only be false?

michaelDCurran commented on source/gui/addonGui.py:
```diff
@@ -206,18 +221,26 @@ def refreshAddonsList(self,activeIndex=0):
-			# Translators: A message asking the user if they wish to restart NVDA as addons have been added or removed. 
-			if gui.messageBox(_("Add-ons have been added or removed. You must restart NVDA for these changes to take effect. Would you like to restart now?"),
+			# Translators: A message asking the user if they wish to restart NVDA as addons have been added, enabled/disabled or removed. 
+			if gui.messageBox(_("Add-ons have been added, enabled, disabled or removed. You must restart NVDA for these changes to take effect. Would you like to restart now?"),
```
I think this message is now getting too complicated. Perhaps something like "Changes to add-ons require NVDA to be restarted. Do you wish to restart NVDA now?".

josephsl replied:
I see. The original intent was to tell users what can be done like what we have now (install/remove). But I can see this could be very complicated, especially when translating. I'll apply this and other UI suggestions. Thanks for catching this.

michaelDCurran commented on source/gui/addonGui.py:
```diff
@@ -248,7 +271,23 @@ def onHelp(self, evt):
+		shouldDisable = self._shouldDisable(addon)
+		if shouldDisable:
+			# Translators: Presented when attempting to disable the selected add-on.
+			if gui.messageBox(_("You are about to disable the selected add-on. If you need to use features provided by this add-on, you need to enable this add-on from add-ons manager. Are you sure you wish to continue and disable this add-on?"),
```
This warning is probably not necessary as the action can be easily reversed.

josephsl replied:
I felt it best to warn users before letting them confirm, similar to the warning shown when users install new add-ons. But I can tell this could become annoying for some...

michaelDCurran commented on user_docs/en/userGuide.t2t:
```diff
@@ -1447,8 +1447,14 @@ NVDA will ask if you really wish to do this.
 As with installing, NVDA must be restarted for the add-on to be fully  removed.
 Until you do, a status of "remove" will be shown for that add-on in the list.
 
+To enable or disable an add-on, select an add-on from the list and press Enable or Disable button.
```
Perhaps:
To disable an add-on, press the disable button.
To enable a previously disabled add-on, press the enable button.

josephsl replied:
Ah yes, thanks for checking it for consistency and length.

michaelDCurran commented on user_docs/en/userGuide.t2t:
```diff
@@ -1447,8 +1447,14 @@ NVDA will ask if you really wish to do this.
 Until you do, a status of "remove" will be shown for that add-on in the list.
 
+To enable or disable an add-on, select an add-on from the list and press Enable or Disable button.
+You can disable an add-on if add-on status indicates it is running or enabled, or enable it if add-on is suspended or disabled.
```
Language correction:  in both cases of "if add-on" should be "if the add-on".

michaelDCurran commented on user_docs/en/userGuide.t2t:
```diff
@@ -1447,8 +1447,14 @@ NVDA will ask if you really wish to do this.
+To enable or disable an add-on, select an add-on from the list and press Enable or Disable button.
+You can disable an add-on if add-on status indicates it is running or enabled, or enable it if add-on is suspended or disabled.
+For each press of enable/disable button, add-on status changes to indicate what will happen when NVDA restarts.
+If you do choose to disable an add-on, be sure to confirm your choice when prompted.
```
Per the suggestion that the warning when disabling is removed, this line can be also removed.

josephsl committed:
Merge branch 'master' into i3090

michaelDCurran commented:
A case which is a bit strange is where you disable an add-on, don't restart, and then re-enable the add-on. This causes the status to change to "enable" and a restart of NVDA is still required. Really, the status should go back to "running" and a restart should not be necessary. I think the easiest way is to change needsRestart into a property which checks to see if any of the "pending" properties are true. Instead of it being set by each action explicitly. Then when enabling an add-on who is pending a disable, the pending disable can be turned off and the status set back to "running".

josephsl committed:
Add-ons manager: do not give a warning message when enabling or disabling add-ons, simplify restart prompt. re #3090.

Code review from Mick Curran (NV Access):
* Restart prompt: quite complicated (agreed). Thus simplify to say that changes were made to add-ons.
* Disable warning: Although it is good to give users warnings similar to the one shown when installing an add-on, it would be better to just disable or enable it and be done with it. Thus no warning from now on.
* User guide: linguistic fixes.

josephsl committed:
Add-ons manager: Just say 'running' when re-enabling an add-on without a restart. re #3090

Consider the following scenario: user opens add-ons manager, disables and add-on, then enables it without a restart. Status columns should say 'running' as opposed to 'enable' (the current one is redundant, as enabling means running).
Caught by Mick Curran (NV Access): NvDA will just say 'runing' for this case.

michaelDCurran commented on source/gui/addonGui.py:
```diff
@@ -60,20 +60,25 @@ def __init__(self,parent):
 		self.helpButton.Bind(wx.EVT_BUTTON,self.onHelp)
 		entryButtonsSizer.Add(self.helpButton)
+		# Translators: The label for a button in Add-ons Manager dialog to enable or disable the selected add-on.
+		self.toggleButton=wx.Button(self,label=_("Disable add-on"))
```
Rename the toggleButton variable to enableDisableButton or something similar. ToggleButton I feel is not clear enough...

josephsl committed:
Add-ons manager: renamed toggleButton to enableDisableButton. re #3090

josephsl commented:
Hi,

Regarding the logic on should disable function: It is much better than saying (I think):
return ((not addon.isDisabled and not addon.isPendingEnable and not addon.isPendingDisable)
or (not addon.isDisabled and addon.isPendingEnable and not addon.isPendingDisable)
or (addon.isDisabled and addon.isPendingEnable and not addon.isPendingDisable))

Basically, there are three cases where this is true:

1. Normal state: Add-on is running (addon.isDisabled says False, nothing added to pending enable and disable sets).
2. About to graduate to running state: add-on is suspended (either globally or via this new patch) and has been added to pending enable set.
3. Change of heart: Add-on is running, was once disabled, and now enabled without closing add-ons manager.

The structure of the truth table is:

1. Check if the add-on should be disabled (this takes priority).
2. If add-on is disabled, check if it's not listed in pending disable/enable sets (change of heart).
3. The second condition is grouped in parentheses to give higher priority).
4. All this is negated.

Hope this helps. Thanks.

josephsl committed:
Add-ons manager: just check pendingDisable once. re #3090.

Caught by Mick Curran: testing shows second check of addon.pendingDisable is actually unnecessary (besides being a performance issue later), hence pendingDisable set wilol be consulted only once.

josephsl committed:
Add-ons manager: refactored how needs restart flag is set by checking for cardinality of sets and comparing sets. re #3090.

As noted by Michael Curran: it is better to make needsRestart a property. However given that it is consulted only when closing, it is better to refactor by:
* Check if any add-ons have been installed or removed ( tehnically, pending).
* Check if disabled add-ons and pending disable set are the same (for the most part, it is).
This also fixes an issue where if restart dialog is dismissed when closing add-ons manager, and add-ons manager is opened again, restart prompt would not be displayed.

josephsl committed:
Add-ons manager and add-ons subsystem: Persistence logic in addonHandler.terminate, rewrote needsRestart. re #3090

Comments from Tyler Spivey and Michael Curran: when disabled, an add-on should stay disabled until enabled later. To hold this informaiton, addonHandler.termiante is now fleshed out with various set operations (intersection nad set difference mostly) to let NVDA know which add-ons should stay deactivated.
Also rewrote needsRestart so it'll go through add-ons and raise the flag if the following conditions are seen: pending install, pending remove, about to come back to life (enabled), and about to stay ou4t of business (disabled).

josephsl committed:
Oops, indentation

michaelDCurran commented on user_docs/en/userGuide.t2t:
```diff
@@ -1447,8 +1447,14 @@ NVDA will ask if you really wish to do this.
+To disable an add-on, press the disable button.
+To enable a previously disabled add-on, press the enable button.
+You can disable an add-on if the add-on status indicates it is running or enabled, or enable it if the add-on is suspended or disabled.
+For each press of enable/disable button, add-on status changes to indicate what will happen when NVDA restarts.
```
Needs a "the": "each press of [the] enable/disable button,"

josephsl committed:
User guide: grammar correction.

Caught by Michael Curran: add 'the' to make it grammatically correct.

josephsl committed:
Add-ons manager: Record names of add-ons that are disabled in hopes that we have a persistent backup in case of disasters such as crashes. re #3090.

A useful that wasn't thought of before (outlined by Michael Curran): what if NVDA suddenly crashes? If this happens, the previous states map would not record add-ons that are indeed disabled because pendingDisableSet is gone. Thus two new sets are introduced:
* DisabledAddons: records add-ons that are actually disabled.
* pendingEnableSet: Tells NVDA to enable the listed add-ons next time it is run.
Due to this change:
* pendingDisableSet will no longer record add-ons that'll be disabled during the next session. This set will now hold a list of add-ons that'll be turned off next time it loads.
* Runtime counterparts of these sets will be kept in memory. When a crash occurs, persistent copy will be loaded.

josephsl committed:
Add-ons Manager/optimization: due to use of states, no need to use runtime states denoting future enable/disable. re #3090

josephsl committed:
Add-ons manager: sshow enabled state for consistency. re #3090

josephsl committed:
Add-ons manager: show enable if and only if add-on is suspended and is pending enable. re #3090

josephsl committed:
Revert "Add-ons manager: show enable if and only if add-on is suspended and is pending enable. re #3090"

This reverts commit f4e3852f707417460c9f44d090ccfe6e4b065668.

josephsl committed:
Cosmetics - addons -> add-ons in core.py

josephsl committed:
Another cosmetics - addon system. -> add-ons system
